### PR TITLE
Extend printk with custom argument getter which is used by the logger

### PR DESCRIPTION
**Pull request by nordic-krch**
Logger has arguments stored in the message. Log output module so far had to read those arguments from message into array and then call vprintk which gets `va_list`. It lead to crazy switch for each number of arguments. By adding custom function for getting arguments (when printk is used) rom is reduced for ~550bytes (on nrf52 board). Stack usage of this function is also reduced because array of arguments is not allocated on the stack.

Note that in future, if floats formatting is added to `z_vprintk`, this approach (modified a bit) can be used to fetch floats from log message.

### CODE REVIEW

File: lib/os/printk.c [hunk 1]
```diff
@@ -281,6 +304,24 @@ void z_vprintk(out_func_t out, void *ctx, const char *fmt, va_list ap)
 	}
 }
 
+void arg_get(void *ctx, void *data, size_t arg_size)
+{
+	va_list *ap = (va_list *)ctx;
+
+	if (arg_size <= sizeof(u32_t)) {
```
Comment (aescolar) [inline lib/os/printk.c]: @nordic-krch I'm afraid this path is a dead end.
This is not portable, you cannot assume you can do this.
That is why we have the va_arg() macros.

Comment (nordic-krch) [reply lib/os/printk.c]: Don't give up so early, can we make it portable then? Is it portable assumption that all types smaller than int are promoted to int? Is it fair to assume that if size is bigger than 32bits then size is 64bit? What if test is written that checks all types. I guess that tests are executed on all qemu platforms we support, right?

Comment (nordic-krch) [reply lib/os/printk.c]: this change saves 550 bytes of rom from being wasted so far and may allow to log floats in the logger (where they are not promoted to double) so i wouldn't like to give up so early.

Comment (aescolar) [reply lib/os/printk.c]: > Is it portable assumption that all types smaller than int are promoted to int? 
> Is it fair to assume that if size is bigger than 32bits then size is 64bit? 

Not really. The calling convention is up to the compiler and platform. And so is the int and pointer size.

> What if test is written that checks all types. I guess that tests are executed on all qemu platforms we support, right?

Well.. the Zephyr kernel is meant to work with any possible platform, not just the ones we have today in CI. Sprinkling assumptions like this in the kernel code would be a long term trap. But I think @andrewboie or @andyross would be best to comment.

**Review by pabigot — CHANGES_REQUESTED** [review]
If this PR is intended to help solve #18351 it should be linked in the description.  I'm not clear on the motivation.

Passing a `va_list *` to a supporting function that invokes `va_arg` with it is portable in principle, but I may be missing a subtlety that @aescolar caught.

I do think the implementation is fragile, and something like this would be better (though it might still technically risk undefined behavior for a pointer argument):

```
union {
  u8_t u8;
  u16_t u16;
  u32_t u32;
  u64_t u64;
} u;

switch (arg_size) {
  case sizeof(u8):
    u.u8 = va_arg(ap, u8_t);
    break;
  // etc ...
  default:
    do_something();
    break;
}
memcpy(data, &u.u8, arg_size);
```

In particular I believe this avoids concerns about unhandled sizes and value alignment that could be problematic with the original proposal.

I would also like clarification on this statement:
> this [...] may allow to log floats in the logger (where they are not promoted to double)

A `float` will always be promoted to `double` when passed into a function through a variadic argument.  I think the explanation is that in the described case macro magic is used to store the float into space required for a `long`, so the `long` is passed instead of a `float`.  I'd be more concerned about undefined behavior in that area.  (My concern at https://github.com/zephyrproject-rtos/zephyr/issues/18351#issuecomment-522276594 about the risk of inferring argument size requirements from the format string is still present, but is also not new.)

File: lib/os/printk.c (hunk 1 above)
Comment (aescolar) [reply lib/os/printk.c]: @nordic-krch Ok, I was probably too fatalistic in my statement. You can make it work portably. But not if you assume that everything smaller than 32bit is pushed as 32bit.
I guess one option that should be fully safe (given that printk only calls with a few types) is to actually handle each type printk will call it with independently. For example by passing to it a "type" identifier (similar to the format modifiers) and switching on it.

Anyhow, it is not me who owns this code, and what you had there would probably work with normal 32 µCs out there (not 16bit ones). So I withdraw from the discussion, and let the area maintainers provide more sensible input.

**Pull request closed without merging**